druid/src/widget/textbox.rs

```diff
@@ -360,71 +369,59 @@ impl Widget<String> for TextBox {
 
             self.do_edit_action(edit_action, data);
             self.reset_cursor_blink(ctx);
-            if data.is_empty() {
-                self.text.set_text(self.placeholder.as_str().into());
-                self.selection = Selection::caret(0);
-            } else {
-                self.text.set_text(data.as_str().into());
-            }
-            self.text.rebuild_if_needed(ctx.text(), env);
 
             if !is_select_all {
                 self.update_hscroll();
             }
         }
     }
 
-    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &String, env: &Env) {
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &String, _env: &Env) {
         match event {
             LifeCycle::WidgetAdded => {
                 ctx.register_for_focus();
-                if data.is_empty() {
-                    self.text.set_text(self.placeholder.as_str().into());
-                    self.text.set_text_color(theme::PLACEHOLDER_COLOR);
-                } else {
-                    self.text.set_text(data.as_str().into());
-                }
-                self.text.rebuild_if_needed(ctx.text(), env);
+                self.text.set_text(data.as_str().into());
             }
             // an open question: should we be able to schedule timers here?
             LifeCycle::FocusChanged(true) => ctx.submit_command(RESET_BLINK.to(ctx.widget_id())),
             _ => (),
         }
     }
 
-    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &String, data: &String, env: &Env) {
-        let content = if data.is_empty() {
-            &self.placeholder
-        } else {
-            data
-        };
-
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &String, data: &String, _env: &Env) {
         // setting text color rebuilds layout, so don't do it if we don't have to
         if !old_data.same(data) {
-            self.selection = self.selection.constrain_to(content);
-            self.text.set_text(content.as_str().into());
-            if data.is_empty() {
-                self.text.set_text_color(theme::PLACEHOLDER_COLOR);
-            } else {
-                self.text.set_text_color(theme::LABEL_COLOR);
-            }
+            self.selection = self.selection.constrain_to(data);
+            self.text.set_text(data.as_str().into());
+            ctx.request_layout();
         }
 
-        self.text.rebuild_if_needed(ctx.text(), env);
-        ctx.request_paint();
+        if ctx.env_changed()
+            && self.placeholder.needs_rebuild_after_update(ctx)
+                | self.text.needs_rebuild_after_update(ctx)
```

Comment:
Just curious why single `|` here? It feels to me like `||` is more idiomatic to indicate "or".